cranelift-filetests/src/function_runner.rs

```diff
@@ -76,14 +324,159 @@ impl FunctionRunner {
         };
 
         let code_page = code_page.make_exec().map_err(|e| e.to_string())?;
-        let callable_fn: fn() -> bool = unsafe { mem::transmute(code_page.as_ptr()) };
 
-        // execute
-        if callable_fn() {
-            Ok(())
-        } else {
-            Err(format!("Failed: {}", context.func.name.to_string()))
+        // Merge two i64 to i128 if necessary.
+        let ret_value_type = return_type(&context.func)?;
+
+        match &self.action {
+            FunctionRunnerAction::Test => {
+                if !ret_value_type.is_bool() {
+                    return Err(format!(
+                        "Function {} return type is not boolean",
+                        context.func.name
+                    ));
+                }
+                if Self::invoke::<bool>(&code_page) {
+                    Ok(())
+                } else {
+                    Err(format!("Failed: {}", context.func.name))
+                }
+            }
+            FunctionRunnerAction::Print => Self::print(code_page, ret_value_type),
+            FunctionRunnerAction::TestWithResult { operator, expected } => {
+                Self::check(code_page, ret_value_type, *operator, *expected)
+            }
+        }
+    }
+}
+
+fn return_type(func: &Function) -> Result<types::Type, String> {
+    let normal_rets: Vec<&AbiParam> = func
+        .signature
+        .returns
+        .iter()
+        .filter(|r| r.purpose == ArgumentPurpose::Normal)
+        .collect();
+    if normal_rets.len() > 2 || normal_rets.len() == 0 {
+        return Err(String::from("Function return type is not supported"));
+    }
+    if normal_rets.len() == 2 {
+        // Only two i64 return arguments supported and they are evaluated as a i128.
+        let both_i64 = normal_rets
+            .into_iter()
+            .map(|p| p.value_type)
+            .all(|ty| ty == types::I64);
```

Comment:
It seems it would be straightforward to check the type of the two array values at once here.